recursion/gnark/sp1/groth16.go

```diff
@@ -66,66 +65,48 @@ func BuildGroth16(buildDir string) error {
 	return nil
 }
 
-// Generate a gnark groth16 proof for a given witness and write the proof to a file. Reads the
-// R1CS, the proving key and the verifier key from the build directory.
-func ProveGroth16(buildDir string, witnessPath string, proofPath string) error {
-	// Read the R1CS.
-	fmt.Println("Reading r1cs...")
-	r1csFile, err := os.Open(buildDir + "/circuit_groth16.bin")
+func ProveVerifyAndSerializeGroth16(witnessInput WitnessInput, r1cs constraint.ConstraintSystem, pk groth16.ProvingKey, vk groth16.VerifyingKey) (Groth16Proof, error) {
+	// Generate the witness.
+	assignment := NewCircuitFromWitness(witnessInput)
+	witness, err := frontend.NewWitness(&assignment, ecc.BN254.ScalarField())
 	if err != nil {
-		return err
+		return Groth16Proof{}, err
 	}
-	r1cs := groth16.NewCS(ecc.BN254)
-	r1cs.ReadFrom(r1csFile)
 
-	// Read the proving key.
-	fmt.Println("Reading pk...")
-	pkFile, err := os.Open(buildDir + "/pk_groth16.bin")
+	// Generate the proof.
+	proof, err := groth16.Prove(r1cs, pk, witness)
```

Comment:
Note: Removed print statements here that were previously used for logging the time to generate the witness & generate the proof. If we want to add this in the future, we should migrate to using a go logger library, which can be added in a separate PR.